opentelemetry/tracing_interceptor.go

```
@@ -0,0 +1,202 @@
+
+	// SpanStarter is a callback to create spans. If not set, this creates normal
+	// OpenTelemetry spans calling Tracer.Start.
+	SpanStarter func(ctx context.Context, t trace.Tracer, spanName string, opts ...trace.SpanStartOption) trace.Span
+}
+
+type spanContextKey struct{}
+
+const defaultHeaderKey = "_tracer-data"
+
+type tracer struct{ options *TracerOptions }
+
+// NewTracer creates a tracer with the given options. Most callers should use
+// NewTracingInterceptor instead.
+func NewTracer(options TracerOptions) (interceptor.Tracer, error) {
+	if options.Tracer == nil {
+		options.Tracer = otel.GetTracerProvider().Tracer("temporal-sdk-go")
+	}
+	if options.TextMapPropagator == nil {
+		options.TextMapPropagator = DefaultTextMapPropagator
+	}
+	if options.SpanContextKey == nil {
+		options.SpanContextKey = spanContextKey{}
+	}
+	if options.HeaderKey == "" {
+		options.HeaderKey = defaultHeaderKey
+	}
+	if options.SpanStarter == nil {
+		options.SpanStarter = func(
+			ctx context.Context,
+			t trace.Tracer,
+			spanName string,
+			opts ...trace.SpanStartOption,
+		) trace.Span {
+			_, span := t.Start(ctx, spanName, opts...)
+			return span
+		}
+	}
+	return &tracer{&options}, nil
+}
+
+// NewTracingInterceptor creates an interceptor for setting on client options
+// that implements OpenTelemetry tracing for workflows.
+func NewTracingInterceptor(options TracerOptions) (interceptor.Interceptor, error) {
+	t, err := NewTracer(options)
+	if err != nil {
+		return nil, err
+	}
+	return interceptor.NewTracingInterceptor(t), nil
+}
+
+func (t *tracer) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: t.options.SpanContextKey,
+		HeaderKey:      t.options.HeaderKey,
+	}
+}
+
+func (t *tracer) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	ctx := trace.SpanContextFromContext(t.options.TextMapPropagator.Extract(context.Background(), textMapCarrier(m)))
+	if !ctx.IsValid() {
+		return nil, nil
```

Shouldn't this return error?

---

or we choose to silently drop the span?

---

There may be no serialized span. There is a difference between an unmarshal failure and a nil span ref because it doesn't exist. I can make this clearer on the `interceptor.Tracer.UnmarshalSpan` interface docs.

---

I had trouble deciding here. With opentracing I do error, so I can error here too.

---

Changed to return error.